### DISCUSSION TITLE
Weird behavior with tab mapping

### DISCUSSION THREAD

**Discussion by stefanos82** [neoclide/coc.nvim#5133] [General]
I have custom mapping for tab key

```vim
" Use Tab to select completion downwards when popup window opened
inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"
" Use Tab to select completion upwards when popup window opened
inoremap <expr> <S-TAB> pumvisible() ? "\<C-p>" : "\<S-TAB>"
inoremap <expr> <cr> pumvisible() ? "\<C-y>" : "\<cr>"
```

and when I have it disabled, that is completely commented out, I get `no mapping found` with `:verbose imap <tab>`, but as soon as I uncomment it, I get 

```vim
i  <Tab>       * coc#pum#visible() ? coc#pum#next(1) : "\<Tab>"                                  
        Last set from ~/.vim/pack/plugins/start/coc.nvim/autoload/coc/ui.vim line 23
```

Without my own version of mapping, I cannot use `vim-go` at all; it does not accept the Enter key to accept the current highlighted / selected word, nor it accepts the tab key of course.

I'm really confused...

I have even tried your own suggestion from README

```vim
" Use tab for trigger completion with characters ahead and navigate
" NOTE: There's always complete item selected by default, you may want to enable
" no select by `"suggest.noselect": true` in your configuration file
" NOTE: Use command ':verbose imap <tab>' to make sure tab is not mapped by
" other plugin before putting this into your config
inoremap <silent><expr> <TAB>
      \ coc#pum#visible() ? coc#pum#next(1) :
      \ CheckBackspace() ? "\<Tab>" :
      \ coc#refresh()
inoremap <expr><S-TAB> coc#pum#visible() ? coc#pum#prev(1) : "\<C-h>"

" Make <CR> to accept selected completion item or notify coc.nvim to format
" <C-g>u breaks current undo, please make your own choice
inoremap <silent><expr> <CR> coc#pum#visible() ? coc#pum#confirm()
                              \: "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"

function! CheckBackspace() abort
  let col = col('.') - 1
  return !col || getline('.')[col - 1]  =~# '\s'
endfunction
```

and obviously does not work with `vim-go` at all; it makes things worse of course.

If there a way to tell `coc.nvim` to allow me use my own custom mapping for Golang only?

**Comment by fannheyward** [comment]
Check `:h coc_disable_mappings_check`

**Comment by stefanos82** [comment]
I forgot to mention that I have tried this approach and of course, my custom mapping works, but this way I lose `coc.nvim`'s auto-completion benefits.

**Reply by fannheyward** [reply]
You need to setup mappings to use coc.nvim functions. As far as I know, vim-go didn't remap the `tab` key, what do you mean _does not work with vim-go at all_? You can use coc.nvim and vim-go both at the same time.

**Comment by stefanos82** [comment]
Let me show you what's the problem I'm facing.

I have the following mapping for tab:

```vim
" Use Tab to select completion downwards when popup window opened
inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"
" Use Tab to select completion upwards when popup window opened
inoremap <expr> <S-TAB> pumvisible() ? "\<C-p>" : "\<S-TAB>"
inoremap <expr> <cr> pumvisible() ? "\<C-y>" : "\<cr>"
```

If I disable this and leave `coc.nvim` as it is configured, auto-completion works with `coc.nvim`, but cannot choose anything either with tab key or Enter.

if I disable `coc.nvim` with `let g:coc_disable_mappings_check = 1` and enable my aforementioned custom tab mapping, it works in `vim-go` as expected, but now the problem switches to `coc.nvim`, meaning the previous problematic `vim-go` behavior takes place in `coc.nvim` instead!

I'm trying to find the golden ratio to make both of them work, with each its own configuration settings.

**Reply by fannheyward** [reply]
Your mapping for tab is invalid for coc.nvim, coc.nvim didn't use vim's default pum but custom floating window, you need to use the mappings from the README or `:h coc-completion-example`.

**Reply by stefanos82** [reply]
I have also tried this; it worked perfectly only with `coc.nvim`, but not with `vim-go` of course, which is a problem for me...

In other words, the following code

```vim
  function! CheckBackspace() abort
    let col = col('.') - 1
    return !col || getline('.')[col - 1]  =~ '\s'
  endfunction

  " Insert <tab> when previous text is space, refresh completion if not.
  inoremap <silent><expr> <TAB>
	\ coc#pum#visible() ? coc#pum#next(1):
	\ CheckBackspace() ? "\<Tab>" :
	\ coc#refresh()
  inoremap <expr><S-TAB> coc#pum#visible() ? coc#pum#prev(1) : "\<C-h>"
```
 combined with

```vim
inoremap <silent><expr> <CR> coc#pum#visible() ? coc#_select_confirm()
				\: "\<C-g>u\<CR>\<c-r>=coc#on_enter()\<CR>"
```

make Enter or Tab not able to work with `vim-go`; if I disable the `<CR>` mapping, the Enter works, but it causes another problem in `coc.nvim`; the Enter key when used to accept the suggested or highlighted keyword, enters a newline instead of staying in same line so I can continue working...

**Reply by fannheyward** [reply]
I still don't get your problem of vim-go. You want to accept suggestions from vim-go?

**Reply by stefanos82** [reply]
No, I want the Tab key and Enter to work with `vim-go`; that's all...

**Reply by fannheyward** [reply]
AFAIK vim-go didn't mapped Enter/Tab, what do you mean _not able to work with vim-go_?

**Reply by stefanos82** [reply]
Tab or Enter key do not work with `vim-go`, because in order to make it work, it needs a custom mapping, which I mentioned above.

Without it, I cannot use it as I want to; also, if I have disabled, it intervenes in `coc.vim`'s behavior.

If I use the `coc-completion` example mappings from above, I have just noticed it devours my memory and makes my system unresponsive for a few seconds.

I have restored my custom mapping and the memory returned back to normal.

I don't know what to say...I'm just baffled!

**Reply by fannheyward** [reply]
To make vim-go work, you setup a custom mapping for it with `inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"`, my question is: why this is needed? vim-go didn't need to setup `tab` to work.

**Reply by fannheyward** [reply]
I don't use vim-go anymore, using coc.nvim + coc-go instead. When I used vim-go before, the mappings were used for go functions, and no need to setup tab/enter mapping.

**Reply by fannheyward** [reply]
As you said, you don't need to accept suggestions from vim-go, do you use coc.nvim with gopls for completion?

**Reply by stefanos82** [reply]
> To make vim-go work, you setup a custom mapping for it with `inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"`, my question is: why this is needed? vim-go didn't need to setup `tab` to work.

Yes, `vim-go` does not need it to work, but I need the `tab` key to complete either the partially typed word or to cycle through autocomplete suggestions.

Also, this custom mapping complements `coc.nvim`, because without it, it does not behave as expected.

I have also tried `coc.nvim`'s `samples from coc-completion-example` and even though it works, it makes my editor completely unresponsive for lots of seconds and my CPU goes crazy; all cores, not just a single one!

Thus, my need to use my own custom mapping.

**Reply by stefanos82** [reply]
> As you said, you don't need to accept suggestions from vim-go, do you use coc.nvim with gopls for completion?

No, I use whatever `vim-go` offers me; the only missing thing is the tab and enter keys to either complete partial keywords or accept the currently highlighted one.

**Reply by stefanos82** [reply]
What's confusing is that with `let g:coc_disable_mappings_check = 1`, hypothetically speaking, I should be able to use the tab key with `coc.nvim` too, meaning my own custom mapping here, because I have disabled its own `coc.nvim` mappings...but it does not work at all; it just inserts a normal tab character (in case 4 spaces)!

**Reply by fannheyward** [reply]
You want to disable coc.nvim's tab mapping, use your own custom mapping, right? Check `:verbose imap <tab>` with `let g:coc_disable_mappings_check = 1`. After you have `coc_disable_mappings_check`, coc.nvim won't set mapping for you.

**Reply by stefanos82** [reply]
...that's what I said above, but my own tab mapping won't work with `coc.nvim` for some reason; meaning, the autocompletion works, yes, but not the tab key; only the arrows can be used to go up and down to highlight the active menu keywords, but with Enter key, it does not select anything.

That's why I use my own custom mapping along with `coc.nvim`.

**Reply by fannheyward** [reply]
What's your `:CocInfo`?

**Reply by stefanos82** [reply]
```
## versions

vim version: VIM - Vi IMproved 9.1 9010496
node version: v20.17.0
coc.nvim version: 0.0.82-c5ffbb3c 2024-08-01 17:40:42 +0800
coc.nvim directory: /home/stefanos/.vim/pack/plugins/start/coc.nvim
term: dumb
platform: linux

## Log of coc.nvim

2024-09-03T14:27:42.234 INFO (pid:8488) [plugin] - coc.nvim initialized with node: v20.17.0 after 161
2024-09-03T14:27:44.065 INFO (pid:8488) [attach] - receive notification: showInfo []
```

**Reply by fannheyward** [reply]
To use coc.nvim, you must use mappings from README or `coc-completion-example` to make coc.nvim to work. Your custom mapping like `inoremap <silent><expr> <TAB> pumvisible() ? "\<C-n>" : "\<TAB>"` won't work.

To use vim-go, you need to use your custom mapping, although I don't know this is must have, as completion suggestions by coc.nvim and vim-go all provided by gopls, the same Go language server.

**Reply by stefanos82** [reply]
As I have stated above to previous thread window(s), the `coc-completion-example` affects my editor's performance; it makes my CPU go crazy for a number of seconds. I simply cannot use my computer until it finishes with whatever it's doing.

That is not the case when I use my custom mapping along with `coc.nvim`.

**Reply by fannheyward** [reply]
Can't reproduce the performance issue. Try to add the mapping from `coc-completion-example` only what you need to your vimrc, not the whole configuration.

**Reply by stefanos82** [reply]
I have and caused the problems I have mentioned.

**Reply by fannheyward** [reply]
Can't reproduce the issue in my tests, try minimal vimrc, only coc.nvim plugin + coc.nvim's mappings.

**Reply by stefanos82** [reply]
Anyway, I will continue with the current setup I have and it will be alright.

Thank you very much.

**Comment by stefanos82** [comment]
@fannheyward Seems like I was very tired to see the right solution to my problem.

I did not need to set a custom mapping; just creating a "bridge" between `coc.nvim` settings and `vim-go` worked flawlessly.

Here's the missing part in `:CocConfig` settings on my side that fixed the problematic behavior:

```json
"golang": {
    "command": "gopls",
    "rootPatterns": ["go.mod", ".git/"],
    "filetypes": ["go"]
}
```

and the necessary `vim-go` flags to match the desired output:

```vim
let g:go_highlight_fields = 1
let g:go_highlight_build_constraints = 1
let g:go_highlight_types = 1
let g:go_highlight_extra_types = 1
let g:go_highlight_format_strings = 1
let g:go_highlight_functions = 1
let g:go_highlight_function_calls = 1
let g:go_highlight_function_parameters = 1
let g:go_highlight_operators = 1
let g:go_highlight_variable_declarations = 1

let g:go_code_completion_icase = 1
" Disable GoDoc
let g:go_doc_keywordprg_enabled = 0
```

This works wonders!
